src/bounces/mod.rs

```diff
@@ -0,0 +1,143 @@
+
+use auth_db::{BounceRecord, BounceType, Db, DbError};
+use settings::{BounceLimit, BounceLimits, Settings};
+
+#[cfg(test)]
+mod test;
+
+#[derive(Debug)]
+pub struct BounceError
+{
+  pub address: String,
+  pub bounce: Option<BounceRecord>,
+  description: String,
+}
+
+impl BounceError
+{
+  pub fn new(address: &str, bounce: BounceRecord) -> BounceError
+  {
+    let description = format!(
+      "email address violated {} limit",
+      match bounce.bounce_type {
+        BounceType::Hard => "hard bounce",
+        BounceType::Soft => "soft bounce",
+        BounceType::Complaint => "complaint",
+      }
+    );
+
+    BounceError {
+      address: address.to_string(),
+      bounce: Some(bounce),
+      description,
+    }
+  }
+}
+
+impl Error for BounceError
+{
+  fn description(&self) -> &str
+  {
+    &self.description
+  }
+}
+
+impl Display for BounceError
+{
+  fn fmt(&self, formatter: &mut Formatter) -> fmt::Result
+  {
+    write!(formatter, "{}", self.description)
+  }
+}
+
+impl From<DbError> for BounceError
+{
+  fn from(error: DbError) -> BounceError
+  {
+    BounceError {
+      address: String::from(""),
+      bounce: None,
+      description: format!("database error: {}", error.description()),
+    }
+  }
+}
+
+impl From<BounceError> for Failure
+{
+  fn from(_error: BounceError) -> Failure
+  {
+    // Eventually we should be able to do something richer than this,
+    // as per https://github.com/SergioBenitez/Rocket/issues/586.
+    Failure(Status::TooManyRequests)
+  }
+}
+
+pub struct Bounces<'a>
+{
+  db: Box<&'a Db>,
+  limits: &'a BounceLimits,
+}
+
+impl<'a> Bounces<'a>
+{
+  pub fn new(settings: &'a Settings, db: Box<&'a Db>) -> Bounces<'a>
+  {
+    Bounces {
+      db,
+      limits: &settings.bouncelimits,
+    }
+  }
+
+  pub fn check(&self, address: &str) -> Result<(), BounceError>
+  {
+    let bounces = self.db.get_email_bounces(address)?;
+    let now = SystemTime::now()
+      .duration_since(SystemTime::UNIX_EPOCH)
+      .expect("system time error");
+    let now = now.as_secs() * 1000;
+    bounces
+      .iter()
+      .try_fold(HashMap::new(), |mut counts, bounce| {
+        {
+          let count = counts.entry(&bounce.bounce_type).or_insert(0);
+          *count += 1;
+          let limits = match bounce.bounce_type {
+            BounceType::Hard => &self.limits.hard,
+            BounceType::Soft => &self.limits.soft,
+            BounceType::Complaint => &self.limits.complaint,
+          };
+          if is_bounce_violation(*count, bounce.created_at, now, limits) {
+            return Err(BounceError::new(address, *bounce));
+          }
+        }
+
+        Ok(counts)
+      })
+      .map(|_| ())
+  }
+}
+
+unsafe impl<'a> Sync for Bounces<'a> {}
+
+fn is_bounce_violation(count: u8, created_at: u64, now: u64, limits: &Vec<BounceLimit>) -> bool
+{
+  for limit in limits.iter() {
+    if count > limit.limit && created_at >= now - limit.period {
+      return true;
```

Comment:
A side-effect of using a data format that maps to strong types is that I find this logic much easier to read than the [equivalent code in the auth server](https://github.com/mozilla/fxa-auth-server/blob/6b58bf9b7c856bfcd9e90f8dcc33e3b0ecaac23f/lib/bounces.js#L58-L69).